File: src/lang/lexer.zig
//! Lexer for tokenizing source code
//!
//! Wraps a generic RegexTrie and associates patterns with TokenTypes
//! for syntax highlighting and parsing.

const std = @import("std");
const regex_trie = @import("regex_trie.zig");
const parse_state = @import("../session/parse_state.zig");

pub const Token = parse_state.Token;
pub const TokenType = parse_state.TokenType;

/// Pattern-to-TokenType mapping entry
pub const TokenPattern = struct {
    pattern: []const u8,
    token_type: TokenType,
};

/// Lexer configuration with pattern definitions
pub const LexerConfig = struct {
    patterns: []const TokenPattern,

    /// Logos language lexer configuration
    /// Order matters: longer/more specific patterns should come first
    /// Keywords/types/builtins use (?![a-zA-Z0-9_]) negative lookahead for word boundaries
    pub fn logosPatterns() LexerConfig {
        // Word boundary: ensures keyword doesn't match if followed by identifier char
        const wb = "(?![a-zA-Z0-9_])";

        const patterns = [_]TokenPattern{
            // ============ Comments (highest priority) ============
            .{ .pattern = "//[^\\n]*", .token_type = .comment },
            .{ .pattern = "/\\*[\\s\\S]*?\\*/", .token_type = .comment },

            // ============ Type names (for casting) ============
            .{ .pattern = "f32" ++ wb, .token_type = .type_name },
            .{ .pattern = "f64" ++ wb, .token_type = .type_name },
            .{ .pattern = "i32" ++ wb, .token_type = .type_name },
            .{ .pattern = "i64" ++ wb, .token_type = .type_name },
            .{ .pattern = "bool" ++ wb, .token_type = .type_name },
            .{ .pattern = "vec2" ++ wb, .token_type = .type_name },
            .{ .pattern = "vec3" ++ wb, .token_type = .type_name },
            .{ .pattern = "vec4" ++ wb, .token_type = .type_name },

            // ============ Keywords ============
            .{ .pattern = "if" ++ wb, .token_type = .keyword },
            .{ .pattern = "else" ++ wb, .token_type = .keyword },
            .{ .pattern = "for" ++ wb, .token_type = .keyword },
            .{ .pattern = "while" ++ wb, .token_type = .keyword },
            .{ .pattern = "and" ++ wb, .token_type = .keyword },
            .{ .pattern = "or" ++ wb, .token_type = .keyword },
            .{ .pattern = "true" ++ wb, .token_type = .keyword },
            .{ .pattern = "false" ++ wb, .token_type = .keyword },

            // ============ Built-in functions ============
            .{ .pattern = "sin" ++ wb, .token_type = .builtin },
            .{ .pattern = "cos" ++ wb, .token_type = .builtin },
            .{ .pattern = "tan" ++ wb, .token_type = .builtin },
            .{ .pattern = "asin" ++ wb, .token_type = .builtin },
            .{ .pattern = "acos" ++ wb, .token_type = .builtin },
            .{ .pattern = "atan" ++ wb, .token_type = .builtin },
            .{ .pattern = "sinh" ++ wb, .token_type = .builtin },
            .{ .pattern = "cosh" ++ wb, .token_type = .builtin },
            .{ .pattern = "tanh" ++ wb, .token_type = .builtin },
            .{ .pattern = "log" ++ wb, .token_type = .builtin },
            .{ .pattern = "log2" ++ wb, .token_type = .builtin },
            .{ .pattern = "log10" ++ wb, .token_type = .builtin },
            .{ .pattern = "exp" ++ wb, .token_type = .builtin },
            .{ .pattern = "exp2" ++ wb, .token_type = .builtin },
            .{ .pattern = "sqrt" ++ wb, .token_type = .builtin },
            .{ .pattern = "pow" ++ wb, .token_type = .builtin },
            .{ .pattern = "abs" ++ wb, .token_type = .builtin },
            .{ .pattern = "sign" ++ wb, .token_type = .builtin },
            .{ .pattern = "floor" ++ wb, .token_type = .builtin },
            .{ .pattern = "ceil" ++ wb, .token_type = .builtin },
            .{ .pattern = "round" ++ wb, .token_type = .builtin },
            .{ .pattern = "fract" ++ wb, .token_type = .builtin },
            .{ .pattern = "mod" ++ wb, .token_type = .builtin },
            .{ .pattern = "min" ++ wb, .token_type = .builtin },
            .{ .pattern = "max" ++ wb, .token_type = .builtin },
            .{ .pattern = "clamp" ++ wb, .token_type = .builtin },
            .{ .pattern = "mix" ++ wb, .token_type = .builtin },
            .{ .pattern = "step" ++ wb, .token_type = .builtin },
            .{ .pattern = "smoothstep" ++ wb, .token_type = .builtin },
            .{ .pattern = "len" ++ wb, .token_type = .builtin },
            .{ .pattern = "length" ++ wb, .token_type = .builtin },
            .{ .pattern = "normalize" ++ wb, .token_type = .builtin },
            .{ .pattern = "dot" ++ wb, .token_type = .builtin },
            .{ .pattern = "cross" ++ wb, .token_type = .builtin },

            // ============ Reserved axis/output variables ============
            .{ .pattern = "axis1" ++ wb, .token_type = .axis },
            .{ .pattern = "axis2" ++ wb, .token_type = .axis },
            .{ .pattern = "axis3" ++ wb, .token_type = .axis },
            .{ .pattern = "time" ++ wb, .token_type = .axis },
            .{ .pattern = "red" ++ wb, .token_type = .axis },
            .{ .pattern = "green" ++ wb, .token_type = .axis },
            .{ .pattern = "blue" ++ wb, .token_type = .axis },
            .{ .pattern = "alpha" ++ wb, .token_type = .axis },

            // ============ Numbers ============
            // Scientific notation
            .{ .pattern = "[0-9]+\\.?[0-9]*[eE][+-]?[0-9]+", .token_type = .number },
            // Float with decimal
            .{ .pattern = "[0-9]+\\.[0-9]+", .token_type = .number },
            // Integer
            .{ .pattern = "[0-9]+", .token_type = .number },

            // ============ Identifiers ============
            // Must come after keywords/builtins to allow them to match first
            .{ .pattern = "[a-zA-Z_][a-zA-Z0-9_]*", .token_type = .identifier },

            // ============ Multi-character operators ============
            .{ .pattern = "!=", .token_type = .operator },
            .{ .pattern = "<=", .token_type = .operator },
            .{ .pattern = ">=", .token_type = .operator },

            // ============ Single-character operators ============
            .{ .pattern = "\\+", .token_type = .operator },
            .{ .pattern = "-", .token_type = .operator },
            .{ .pattern = "\\*", .token_type = .operator },
            .{ .pattern = "/", .token_type = .operator },
            .{ .pattern = "%", .token_type = .operator },
            .{ .pattern = "\\^", .token_type = .operator },
            .{ .pattern = "<", .token_type = .operator },
            .{ .pattern = ">", .token_type = .operator },
            .{ .pattern = "=", .token_type = .operator },
            .{ .pattern = "!", .token_type = .operator },

            // Unicode superscript for square (²)
            .{ .pattern = "\xc2\xb2", .token_type = .operator }, // UTF-8 encoding of ²

            // ============ Punctuation ============
            .{ .pattern = "\\(", .token_type = .punctuation },
            .{ .pattern = "\\)", .token_type = .punctuation },
            .{ .pattern = "\\[", .token_type = .punctuation },
            .{ .pattern = "\\]", .token_type = .punctuation },
            .{ .pattern = ",", .token_type = .punctuation },
            .{ .pattern = ":", .token_type = .punctuation },
            .{ .pattern = "\\.", .token_type = .punctuation },

            // ============ Whitespace ============
            .{ .pattern = "[ \\t\\n\\r]+", .token_type = .whitespace },
        };
        return .{ .patterns = &patterns };
    }

    /// Legacy/default lexer configuration (generic)
    pub fn syntaxTokenPatterns() LexerConfig {
        return logosPatterns();
    }
};

/// Lexer wrapping a RegexTrie
pub const Lexer = struct {
    trie: *regex_trie.RegexTrie,
    /// Token type storage (owned by lexer, pointed to by RegexTrieValue.data)
    token_types: std.ArrayList(TokenType),
    allocator: std.mem.Allocator,

    /// Initialize a lexer with the given configuration
    pub fn init(allocator: std.mem.Allocator, config: LexerConfig) !Lexer {
        var trie = try regex_trie.RegexTrie.init(allocator);
        errdefer trie.deinit();

        var token_types = try std.ArrayList(TokenType).initCapacity(allocator, config.patterns.len);
        errdefer token_types.deinit(allocator);

        // Insert all patterns into the trie with TokenType as data
        for (config.patterns) |pattern_entry| {
            // Store the token type in our list
            try token_types.append(allocator, pattern_entry.token_type);
            const token_type_ptr = &token_types.items[token_types.items.len - 1];

            // Insert with pointer to the token type as data
            try trie.insert(allocator, pattern_entry.pattern, @ptrCast(token_type_ptr));
        }

        return .{
            .trie = trie,
            .token_types = token_types,
            .allocator = allocator,
        };
    }

    /// Deinitialize the lexer
    pub fn deinit(self: *Lexer) void {
        self.trie.deinit();
        self.token_types.deinit(self.allocator);
    }

    /// Tokenize input text into a token array
    /// Caller owns the returned slice and must free it
    pub fn tokenize(self: *Lexer, content: []const u8) ![]Token {
        var tokens = try std.ArrayList(Token).initCapacity(self.allocator, 0);
        errdefer tokens.deinit(self.allocator);

        var pos: usize = 0;
        while (pos < content.len) {
            const remaining = content[pos..];

            // Try to match a token
            const match_result = self.trie.get(remaining) catch |err| {
                if (err == regex_trie.RegexTrieError.NodeNotFound) {
                    // No match - consume the full UTF-8 codepoint (not just 1 byte)
                    // to avoid splitting multi-byte characters like π (0xCF 0x80) across tokens.
                    const byte_len = std.unicode.utf8ByteSequenceLength(content[pos]) catch 1;
                    const end = @min(pos + byte_len, content.len);
                    try tokens.append(self.allocator, .{
                        .text = content[pos..end],
                        .token_type = .unknown,
                        .byte_start = pos,
                        .byte_end = end,
                    });
                    pos = end;
                    continue;
                }
                return err;
            };

            // Extract TokenType from the data field
            const token_type = if (match_result.data) |data_ptr|
                @as(*TokenType, @ptrCast(@alignCast(data_ptr))).*
            else
                .unknown;

            const matched_len = match_result.matched;

            try tokens.append(self.allocator, .{
                .text = content[pos .. pos + matched_len],
                .token_type = token_type,
                .byte_start = pos,
                .byte_end = pos + matched_len,
            });

            pos += matched_len;
        }

        return try tokens.toOwnedSlice(self.allocator);
    }

    /// Check if content needs re-lexing by comparing with cached hash
    pub fn needsRelex(state: *parse_state.ParseState, content: []const u8) bool {
        return state.shouldRelex(content);
    }
};

File: src/lang/lexer.rs
use super::token::{Token, TokenType};

pub(crate) const BUILTINS: &[&str] = &[
    "sin", "cos", "tan", "asin", "acos", "atan", "sinh", "cosh", "tanh",
    "log", "log2", "log10", "exp", "exp2",
    "floor", "ceil", "round", "fract",
    "abs", "sign", "pow", "sqrt", "mod", "min", "max",
    "clamp", "mix", "step", "smoothstep",
    "length", "normalize", "dot", "cross",
];

pub(crate) const AXIS_VARS: &[&str] = &["x", "y", "z", "time"];

pub(crate) const TYPE_NAMES: &[(&str, TokenType)] = &[
    ("f32", TokenType::TypeF32),
    ("f64", TokenType::TypeF64),
    ("i32", TokenType::TypeI32),
    ("vec2", TokenType::TypeVec2),
    ("vec3", TokenType::TypeVec3),
    ("vec4", TokenType::TypeVec4),
    ("bool", TokenType::TypeBool),
];

pub struct Lexer<'a> {
    input: &'a str,
    pos: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self { input, pos: 0 }
    }

    pub fn tokenize(&mut self) -> Result<Vec<Token>, String> {
        let mut tokens = Vec::new();
        loop {
            self.skip_spaces_and_comments();
            if self.pos >= self.input.len() {
                tokens.push(Token { ty: TokenType::Eof, span: (self.pos, self.pos) });
                break;
            }
            // Newlines are statement separators
            if self.peek() == Some('\n') {
                let start = self.pos;
                // Collapse consecutive newlines into one
                while self.peek() == Some('\n') {
                    self.advance();
                    self.skip_spaces_and_comments();
                }
                // Only emit newline if there are already tokens (skip leading newlines)
                // and we're not at EOF
                if !tokens.is_empty() && self.pos < self.input.len() {
                    tokens.push(Token { ty: TokenType::Newline, span: (start, self.pos) });
                }
                continue;
            }
            let tok = self.next_token()?;
            tokens.push(tok);
        }
        Ok(tokens)
    }

    /// Skip spaces, tabs, and comments — but NOT newlines.
    fn skip_spaces_and_comments(&mut self) {
        loop {
            // Skip non-newline whitespace
            while self.pos < self.input.len() {
                match self.peek() {
                    Some(c) if c.is_whitespace() && c != '\n' => {
                        self.pos += c.len_utf8();
                    }
                    _ => break,
                }
            }
            // Skip line comments (consume up to but not including newline)
            if self.starts_with("//") {
                while self.pos < self.input.len() && self.peek() != Some('\n') {
                    self.pos += self.peek().unwrap().len_utf8();
                }
                continue;
            }
            // Skip block comments
            if self.starts_with("/*") {
                self.pos += 2;
                while self.pos < self.input.len() && !self.starts_with("*/") {
                    self.pos += self.peek().unwrap().len_utf8();
                }
                if self.starts_with("*/") {
                    self.pos += 2;
                }
                continue;
            }
            break;
        }
    }

    fn next_token(&mut self) -> Result<Token, String> {
        let start = self.pos;
        let ch = self.peek().unwrap();

        // Two-character operators (must check before single-char)
        if let Some(ty) = self.try_two_char_op() {
            return Ok(Token { ty, span: (start, self.pos) });
        }

        // Single-character tokens
        match ch {
            '+' => { self.advance(); return Ok(Token { ty: TokenType::Plus, span: (start, self.pos) }); }
            '-' => { self.advance(); return Ok(Token { ty: TokenType::Minus, span: (start, self.pos) }); }
            '*' => { self.advance(); return Ok(Token { ty: TokenType::Star, span: (start, self.pos) }); }
            '/' => { self.advance(); return Ok(Token { ty: TokenType::Slash, span: (start, self.pos) }); }
            '%' => { self.advance(); return Ok(Token { ty: TokenType::Percent, span: (start, self.pos) }); }
            '^' => { self.advance(); return Ok(Token { ty: TokenType::Caret, span: (start, self.pos) }); }
            '<' => { self.advance(); return Ok(Token { ty: TokenType::Lt, span: (start, self.pos) }); }
            '>' => { self.advance(); return Ok(Token { ty: TokenType::Gt, span: (start, self.pos) }); }
            // `=` is equality in Logos (not assignment)
            '=' => { self.advance(); return Ok(Token { ty: TokenType::Eq, span: (start, self.pos) }); }
            ':' => { self.advance(); return Ok(Token { ty: TokenType::Colon, span: (start, self.pos) }); }
            '(' => { self.advance(); return Ok(Token { ty: TokenType::LParen, span: (start, self.pos) }); }
            ')' => { self.advance(); return Ok(Token { ty: TokenType::RParen, span: (start, self.pos) }); }
            '[' => { self.advance(); return Ok(Token { ty: TokenType::LBracket, span: (start, self.pos) }); }
            ']' => { self.advance(); return Ok(Token { ty: TokenType::RBracket, span: (start, self.pos) }); }
            ',' => { self.advance(); return Ok(Token { ty: TokenType::Comma, span: (start, self.pos) }); }
            '.' => { self.advance(); return Ok(Token { ty: TokenType::Dot, span: (start, self.pos) }); }
            _ => {}
        }

        // Numbers
        if ch.is_ascii_digit() || (ch == '.' && self.peek_at(1).map_or(false, |c| c.is_ascii_digit())) {
            return self.lex_number(start);
        }

        // Unicode math symbols — must check BEFORE identifiers since some
        // (like π, τ) are alphabetic in Unicode but should be their own tokens.
        if ch == '\u{00B2}' { // ² superscript
            self.advance();
            return Ok(Token { ty: TokenType::Builtin("square".to_string()), span: (start, self.pos) });
        }
        if ch == '\u{00B3}' { // ³ superscript
            self.advance();
            return Ok(Token { ty: TokenType::Builtin("cube".to_string()), span: (start, self.pos) });
        }
        if ch == '\u{03C0}' { // π
            self.advance();
            return Ok(Token { ty: TokenType::Number(std::f64::consts::PI), span: (start, self.pos) });
        }
        if ch == '\u{03C4}' { // τ (tau)
            self.advance();
            return Ok(Token { ty: TokenType::Number(std::f64::consts::TAU), span: (start, self.pos) });
        }
        if ch == '\u{2212}' { // − (unicode minus)
            self.advance();
            return Ok(Token { ty: TokenType::Minus, span: (start, self.pos) });
        }
        if ch == '\u{00D7}' { // × (multiplication sign)
            self.advance();
            return Ok(Token { ty: TokenType::Star, span: (start, self.pos) });
        }
        if ch == '\u{00F7}' { // ÷ (division sign)
            self.advance();
            return Ok(Token { ty: TokenType::Slash, span: (start, self.pos) });
        }
        if ch == '\u{222B}' { // ∫ (integral)
            self.advance();
            return Ok(Token { ty: TokenType::Builtin("int".to_string()), span: (start, self.pos) });
        }
        if ch == '\u{2202}' { // ∂ (partial derivative)
            self.advance();
            return Ok(Token { ty: TokenType::Builtin("df".to_string()), span: (start, self.pos) });
        }
        if ch == '\u{2207}' { // ∇ (nabla/gradient)
            self.advance();
            return Ok(Token { ty: TokenType::Identifier("nabla".to_string()), span: (start, self.pos) });
        }
        if ch == '\u{221A}' { // √ (square root)
            self.advance();
            return Ok(Token { ty: TokenType::Builtin("sqrt".to_string()), span: (start, self.pos) });
        }
        if ch == '\u{221E}' { // ∞ (infinity)
            self.advance();
            return Ok(Token { ty: TokenType::Identifier("infinity".to_string()), span: (start, self.pos) });
        }
        if ch == '\u{2211}' { // ∑ (summation)
            self.advance();
            return Ok(Token { ty: TokenType::Builtin("sum".to_string()), span: (start, self.pos) });
        }
        if ch == '\u{220F}' { // ∏ (product)
            self.advance();
            return Ok(Token { ty: TokenType::Builtin("prod".to_string()), span: (start, self.pos) });
        }
        if ch == '\u{2264}' { // ≤
            self.advance();
            return Ok(Token { ty: TokenType::Lte, span: (start, self.pos) });
        }
        if ch == '\u{2265}' { // ≥
            self.advance();
            return Ok(Token { ty: TokenType::Gte, span: (start, self.pos) });
        }
        if ch == '\u{2260}' { // ≠
            self.advance();
            return Ok(Token { ty: TokenType::Neq, span: (start, self.pos) });
        }
        // Superscript digits 0,1,4-9
        if let Some(exp) = match ch {
            '\u{2070}' => Some("0"), '\u{00B9}' => Some("1"),
            '\u{2074}' => Some("4"), '\u{2075}' => Some("5"),
            '\u{2076}' => Some("6"), '\u{2077}' => Some("7"),
            '\u{2078}' => Some("8"), '\u{2079}' => Some("9"),
            _ => None,
        } {
            self.advance();
            return Ok(Token {
                ty: TokenType::Builtin(format!("pow{}", exp)),
                span: (start, self.pos),
            });
        }

        // Identifiers / keywords / builtins
        if ch.is_alphabetic() || ch == '_' {
            return Ok(self.lex_identifier(start));
        }

        // Backslash: skip it gracefully (user typing LaTeX before autocomplete accepts)
        if ch == '\\' {
            self.advance();
            return Ok(Token { ty: TokenType::Identifier("\\".to_string()), span: (start, self.pos) });
        }

        Err(format!("Unexpected character '{}' at position {}", ch, start))
    }

    fn try_two_char_op(&mut self) -> Option<TokenType> {
        // `==` is also equality (same as `=` in Logos, for backward compat)
        if self.starts_with("==") { self.pos += 2; return Some(TokenType::Eq); }
        if self.starts_with("!=") { self.pos += 2; return Some(TokenType::Neq); }
        if self.starts_with("<=") { self.pos += 2; return Some(TokenType::Lte); }
        if self.starts_with(">=") { self.pos += 2; return Some(TokenType::Gte); }
        None
    }

    fn lex_number(&mut self, start: usize) -> Result<Token, String> {
        // Integer part
        while self.pos < self.input.len() && self.peek().map_or(false, |c| c.is_ascii_digit()) {
            self.advance();
        }
        // Decimal part
        if self.peek() == Some('.') && self.peek_at(1).map_or(false, |c| c.is_ascii_digit()) {
            self.advance(); // consume '.'
            while self.pos < self.input.len() && self.peek().map_or(false, |c| c.is_ascii_digit()) {
                self.advance();
            }
        }
        // Scientific notation
        if self.peek().map_or(false, |c| c == 'e' || c == 'E') {
            self.advance();
            if self.peek() == Some('+') || self.peek() == Some('-') {
                self.advance();
            }
            while self.pos < self.input.len() && self.peek().map_or(false, |c| c.is_ascii_digit()) {
                self.advance();
            }
        }

        let text = &self.input[start..self.pos];
        let value: f64 = text.parse().map_err(|e| format!("Invalid number '{}': {}", text, e))?;
        Ok(Token { ty: TokenType::Number(value), span: (start, self.pos) })
    }

    fn lex_identifier(&mut self, start: usize) -> Token {
        while self.pos < self.input.len() && self.peek().map_or(false, |c| {
            (c.is_alphanumeric() || c == '_') && !is_unicode_math_symbol(c)
        }) {
            self.advance();
        }
        let text = &self.input[start..self.pos];

        // Keywords
        let ty = match text {
            "if" => TokenType::If,
            "else" => TokenType::Else,
            "for" => TokenType::For,
            "while" => TokenType::While,
            "and" => TokenType::And,
            "or" => TokenType::Or,
            "not" => TokenType::Not,
            "true" => TokenType::BoolLit(true),
            "false" => TokenType::BoolLit(false),
            _ => {
                // Type names
                for &(name, ref tok) in TYPE_NAMES {
                    if text == name {
                        return Token { ty: tok.clone(), span: (start, self.pos) };
                    }
                }
                // Builtins
                if BUILTINS.contains(&text) {
                    TokenType::Builtin(text.to_string())
                }
                // Axis variables
                else if AXIS_VARS.contains(&text) {
                    TokenType::AxisVar(text.to_string())
                }
                // Regular identifier
                else {
                    TokenType::Identifier(text.to_string())
                }
            }
        };

        Token { ty, span: (start, self.pos) }
    }

    fn peek(&self) -> Option<char> {
        self.input[self.pos..].chars().next()
    }

    fn peek_at(&self, offset: usize) -> Option<char> {
        let pos = self.pos + offset;
        if pos < self.input.len() {
            self.input[pos..].chars().next()
        } else {
            None
        }
    }

    fn advance(&mut self) {
        if let Some(ch) = self.peek() {
            self.pos += ch.len_utf8();
        }
    }

    fn starts_with(&self, s: &str) -> bool {
        self.input[self.pos..].starts_with(s)
    }
}

/// Unicode math symbols that look alphanumeric but should be separate tokens.
/// These must NOT be consumed as part of identifiers.
fn is_unicode_math_symbol(c: char) -> bool {
    matches!(c,
        '\u{00B2}'   // ² superscript 2
        | '\u{00B3}' // ³ superscript 3
        | '\u{03C0}' // π pi
        | '\u{03C4}' // τ tau
        | '\u{2212}' // − unicode minus
        | '\u{00D7}' // × multiplication sign
        | '\u{00F7}' // ÷ division sign
        | '\u{222B}' // ∫ integral
        | '\u{2202}' // ∂ partial
        | '\u{2207}' // ∇ nabla
        | '\u{221A}' // √ sqrt
        | '\u{221E}' // ∞ infinity
        | '\u{2211}' // ∑ sum
        | '\u{220F}' // ∏ product
        | '\u{2264}' // ≤
        | '\u{2265}' // ≥
        | '\u{2260}' // ≠
        | '\u{2070}' // ⁰
        | '\u{00B9}' // ¹
        | '\u{2074}' // ⁴
        | '\u{2075}' // ⁵
        | '\u{2076}' // ⁶
        | '\u{2077}' // ⁷
        | '\u{2078}' // ⁸
        | '\u{2079}' // ⁹
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_tokens() {
        let mut lexer = Lexer::new("x + 3.14 * sin(y)");
        let tokens = lexer.tokenize().unwrap();
        assert!(matches!(tokens[0].ty, TokenType::AxisVar(ref s) if s == "x"));
        assert_eq!(tokens[1].ty, TokenType::Plus);
        assert!(matches!(tokens[2].ty, TokenType::Number(n) if (n - 3.14).abs() < 1e-10));
        assert_eq!(tokens[3].ty, TokenType::Star);
        assert!(matches!(tokens[4].ty, TokenType::Builtin(ref s) if s == "sin"));
        assert_eq!(tokens[5].ty, TokenType::LParen);
        assert!(matches!(tokens[6].ty, TokenType::AxisVar(ref s) if s == "y"));
        assert_eq!(tokens[7].ty, TokenType::RParen);
        assert_eq!(tokens[8].ty, TokenType::Eof);
    }

    #[test]
    fn test_equality_is_eq() {
        // In Logos, `=` is equality, not assignment
        let mut lexer = Lexer::new("a = b");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[1].ty, TokenType::Eq);
    }

    #[test]
    fn test_double_equals_also_eq() {
        let mut lexer = Lexer::new("a == b");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[1].ty, TokenType::Eq);
    }

    #[test]
    fn test_colon_binding() {
        let mut lexer = Lexer::new("r: 5");
        let tokens = lexer.tokenize().unwrap();
        assert!(matches!(tokens[0].ty, TokenType::Identifier(ref s) if s == "r"));
        assert_eq!(tokens[1].ty, TokenType::Colon);
        assert!(matches!(tokens[2].ty, TokenType::Number(n) if (n - 5.0).abs() < 1e-10));
    }

    #[test]
    fn test_comparison_ops() {
        let mut lexer = Lexer::new("a != c <= d >= e");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[1].ty, TokenType::Neq);
        assert_eq!(tokens[3].ty, TokenType::Lte);
        assert_eq!(tokens[5].ty, TokenType::Gte);
    }

    #[test]
    fn test_keywords() {
        let mut lexer = Lexer::new("if true else false and or not");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens[0].ty, TokenType::If);
        assert_eq!(tokens[1].ty, TokenType::BoolLit(true));
        assert_eq!(tokens[2].ty, TokenType::Else);
        assert_eq!(tokens[3].ty, TokenType::BoolLit(false));
        assert_eq!(tokens[4].ty, TokenType::And);
        assert_eq!(tokens[5].ty, TokenType::Or);
        assert_eq!(tokens[6].ty, TokenType::Not);
    }

    #[test]
    fn test_scientific_notation() {
        let mut lexer = Lexer::new("1e5 2.5e-3");
        let tokens = lexer.tokenize().unwrap();
        assert!(matches!(tokens[0].ty, TokenType::Number(n) if (n - 1e5).abs() < 1.0));
        assert!(matches!(tokens[1].ty, TokenType::Number(n) if (n - 2.5e-3).abs() < 1e-10));
    }

    #[test]
    fn test_comments() {
        let mut lexer = Lexer::new("x // comment\n+ y /* block */ * z");
        let tokens = lexer.tokenize().unwrap();
        assert!(matches!(tokens[0].ty, TokenType::AxisVar(ref s) if s == "x"));
        // After comment, there's a newline token, then + y * z
        assert_eq!(tokens[1].ty, TokenType::Newline);
        assert_eq!(tokens[2].ty, TokenType::Plus);
        assert!(matches!(tokens[3].ty, TokenType::AxisVar(ref s) if s == "y"));
        assert_eq!(tokens[4].ty, TokenType::Star);
        assert!(matches!(tokens[5].ty, TokenType::AxisVar(ref s) if s == "z"));
    }

    #[test]
    fn test_newlines_as_separators() {
        let mut lexer = Lexer::new("a: 5\nb: 10\na + b");
        let tokens = lexer.tokenize().unwrap();
        // a : 5 \n b : 10 \n a + b EOF
        assert!(matches!(tokens[0].ty, TokenType::Identifier(ref s) if s == "a"));
        assert_eq!(tokens[1].ty, TokenType::Colon);
        assert!(matches!(tokens[2].ty, TokenType::Number(_)));
        assert_eq!(tokens[3].ty, TokenType::Newline);
        assert!(matches!(tokens[4].ty, TokenType::Identifier(ref s) if s == "b"));
    }

    #[test]
    fn test_empty_input() {
        let mut lexer = Lexer::new("");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].ty, TokenType::Eof);
    }

    #[test]
    fn test_whitespace_only() {
        let mut lexer = Lexer::new("   \t\t  ");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].ty, TokenType::Eof);
    }

    #[test]
    fn test_unicode_superscript_not_part_of_identifier() {
        // ² is alphanumeric in Unicode but must be a separate token, not part of "x"
        let mut lexer = Lexer::new("x\u{00B2}");
        let tokens = lexer.tokenize().unwrap();
        assert!(matches!(tokens[0].ty, TokenType::AxisVar(ref s) if s == "x"),
            "x should be AxisVar, got {:?}", tokens[0].ty);
        assert!(matches!(tokens[1].ty, TokenType::Builtin(ref s) if s == "square"),
            "² should be Builtin(square), got {:?}", tokens[1].ty);
    }

    #[test]
    fn test_unicode_pi_not_part_of_identifier() {
        // π after an identifier should be a separate Number token
        let mut lexer = Lexer::new("r\u{03C0}");
        let tokens = lexer.tokenize().unwrap();
        assert!(matches!(tokens[0].ty, TokenType::Identifier(ref s) if s == "r"));
        assert!(matches!(tokens[1].ty, TokenType::Number(_)));
    }

    #[test]
    fn test_complex_unicode_expression() {
        // x²*y²+sin(x)³-sin(y²)²=9
        let input = "x\u{00B2}*y\u{00B2}+sin(x)\u{00B3}-sin(y\u{00B2})\u{00B2}=9";
        let mut lexer = Lexer::new(input);
        let tokens = lexer.tokenize().unwrap();

        // Expected: x ² * y ² + sin ( x ) ³ - sin ( y ² ) ² = 9 EOF
        let expected: Vec<&str> = vec![
            "AxisVar(x)", "Builtin(square)", "Star",
            "AxisVar(y)", "Builtin(square)", "Plus",
            "Builtin(sin)", "LParen", "AxisVar(x)", "RParen", "Builtin(cube)", "Minus",
            "Builtin(sin)", "LParen", "AxisVar(y)", "Builtin(square)", "RParen", "Builtin(square)",
            "Eq", "Number(9)", "Eof",
        ];

        assert_eq!(
            tokens.len(), expected.len(),
            "Expected {} tokens, got {}: {:?}", expected.len(), tokens.len(),
            tokens.iter().map(|t| format!("{:?}", t.ty)).collect::<Vec<_>>()
        );

        // Verify key tokens
        assert!(matches!(tokens[0].ty, TokenType::AxisVar(ref s) if s == "x"));
        assert!(matches!(tokens[1].ty, TokenType::Builtin(ref s) if s == "square"));
        assert_eq!(tokens[2].ty, TokenType::Star);
        assert!(matches!(tokens[3].ty, TokenType::AxisVar(ref s) if s == "y"));
        assert!(matches!(tokens[4].ty, TokenType::Builtin(ref s) if s == "square"));
        assert_eq!(tokens[5].ty, TokenType::Plus);
        assert!(matches!(tokens[6].ty, TokenType::Builtin(ref s) if s == "sin"));
        assert_eq!(tokens[7].ty, TokenType::LParen);
        assert!(matches!(tokens[8].ty, TokenType::AxisVar(ref s) if s == "x"));
        assert_eq!(tokens[9].ty, TokenType::RParen);
        assert!(matches!(tokens[10].ty, TokenType::Builtin(ref s) if s == "cube"));
        assert_eq!(tokens[11].ty, TokenType::Minus);
        assert!(matches!(tokens[12].ty, TokenType::Builtin(ref s) if s == "sin"));
        assert_eq!(tokens[13].ty, TokenType::LParen);
        assert!(matches!(tokens[14].ty, TokenType::AxisVar(ref s) if s == "y"));
        assert!(matches!(tokens[15].ty, TokenType::Builtin(ref s) if s == "square"));
        assert_eq!(tokens[16].ty, TokenType::RParen);
        assert!(matches!(tokens[17].ty, TokenType::Builtin(ref s) if s == "square"));
        assert_eq!(tokens[18].ty, TokenType::Eq);
        assert!(matches!(tokens[19].ty, TokenType::Number(n) if (n - 9.0).abs() < 1e-10));
    }
}
